soccer/ui/MainWindow.hpp

```diff
@@ -263,4 +249,7 @@ private Q_SLOTS:
     std::vector<QComboBox*> _robotConfigQComboBoxes{};
 
     std::vector<QComboBox*> _robotDebugResponseQComboBoxes{};
+
+    std::mutex _context_mutex;
+    Context* _context;
```

Comment:
perhaps consider making this a `shared_ptr`?

Reply:
Raw pointers are still useful to represent non-owning pointers (depending on who you ask, you could also reasonably use a reference here). This is not an "a owns b" relationship, so it makes sense for it to be a regular raw pointer.

Reply:
Ah, makes sense. Have you considered `weak_ptr`? This seems like the textbook use-case for it (it definitely adds some overhead having to check that the pointer is valid, but in general it's good practice to check that a non-owning pointer is valid before using it).

Reply:
In this case though `Context` is something that basically is guaranteed to outlive `MainWindow` itself, so it doesn't _really_ matter that much I think.

Reply:
Yeah. Personally I consider using weak_ptr/shared_ptr when lifetimes can be statically determined to be a bit of an antipattern. In this case, Context is guaranteed to outlive all of the modules holding a pointer to it.

Reply:
Can it be statically (ie. compile-time) determined that `_context` outlives `MainWindow`? I suppose it can by looking at the code, but if that were ever to change, there'd certainly be no compile-time indication of it, and the only runtime indication would be random segfaults (if we're lucky) or the program somehow misbehaves (much worse).

Totally not here to tell ya'll how to handle this sort of stuff, just a suggestion. Especially for code for student teams I tend to be _very_ conservative in terms of making code as safe as possible.

Reply:
Context lives for the duration of the program (as does MainWindow, but it is destructed just before Context). IMO non-owning pointers only make sense to use in two situations:

 - The pointer is passed into a function, used in that function, and never stored anywhere.
 - The pointer points to an object that has whole-program lifetime (i.e. `'static` in Rust).

In these two situations, there's no ambiguity and so it's acceptable. In addition, IMO it is more clear than a reference (for example, it forces you to make it explicit that the function is allowed to mutate a given parameter). The reason I don't like `shared_`/`weak_ptr` is that they tend to act as a crutch to avoid having to think about ownership, which I've seen lead to messy messy spaghetti code (we've got a fair bit of that in this repository that we're currently trying to clean out). However, they are obviously very useful tools when you actually need any sort of dynamic lifetime.